### PR TITLE
Improve documentation for continuous integration

### PR DESCRIPTION
**Pull request by farski**

### CODE REVIEW

File: ci/CONTRIBUTING.MD [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+- [AWS CodeBuild Build Phase Transitions](https://docs.aws.amazon.com/codebuild/latest/userguide/view-build-details.html#view-build-details-phases)
+- [AWS CodeBuild Build Environment Reference](https://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref.html)
+- [AWS CodeBuild Environment Variables in Build Environments ](https://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref-env-vars.html)
+
+## Deployment
+
+The continuous integration platform is intended to be managed and launched through [AWS CloudFormation](https://aws.amazon.com/cloudformation/), so any AWS resources that are necessary for the system to operate should be added to the `ci.yml` template that already exists. CloudFormation stacks should always be designed to work in any AWS region, and allow for multiple stacks to be launched without conflicting, even in the same region (though doing so is unexpected and not recommended).
+
+**Never make changes changes to AWS resources that were created through CloudFormation, and do not create any resources outside of the CloudFormation template**. The CI system is entirely self-contained and should remain that way.
```
Comment (gcampo88) [inline ci/CONTRIBUTING.MD]: changes changes!

Comment (gcampo88) [reply ci/CONTRIBUTING.MD]: also, do you mean that people shouldn't be making changes / creating new resources _other than through CloudFormation_? (Right now it reads like you shouldn't make any changes ever)

File: ci/CONTRIBUTING.MD [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+When there are external dependencies, such as the configuration of GitHub Webhooks, the [README](https://github.com/PRX/Infrastructure/blob/master/ci/README.md) should include information to help with getting those dependencies set up.
+
+There is no automated process for deploying changes to the CI stack. If the CloudFormation template changes, the stack must be updated manually (through the AWS Console or command line). You should always be sure to carefully review the resources that are changing before executing the stack update. The code for AWS Lambda functions likewise will not deploy automatically, even if you deploy stack changes manually. Even if you push updated Lambda code to S3 prior to executing a stack update, CloudFormation may not always notice the changes. For now, the easiest and most reliable way to deploy Lambda code changes is copy-and-paste through the console. (This is not true of other properties of Lambda functions defined in the template, such as the timeout; those will update as expected during a stack update.)
```
Comment (gcampo88) [inline ci/CONTRIBUTING.MD]: any guidance for what steps to take as part of how to "carefully review the resources that are changing before executing the stack update" ? or any must-do steps?

Comment (gcampo88) [reply ci/CONTRIBUTING.MD]: also would be good to add a little more clarity around Lambda deploy -- easiest + best way currently is to copy + paste through console; do we _also_ need to push updated code to s3? and/or update the s3 code object version ID in secrets anywhere?

File: ci/CONTRIBUTING.MD [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+It is still a good idea to have code reviewed, even if the changes have already been deployed and appear to be working. If you're working on feature branches and deploying code that isn't yet committed to `master`, there is the potential that others are also deploying changes that don't include your changes. Good communication is the best method for handling this.
+
+## Code Standards
+
+Be sure to follow any coding standards outlined in the [CONTRIBUTING](https://github.com/PRX/Infrastructure/blob/master/CONTRIBUTING.md) file in the root of this repository.
+
+## Documentation
+
+The [README](https://github.com/PRX/Infrastructure/blob/master/ci/README.md) should always be kept up to date. As you make changes to parts of the system, make sure you are simultaneously updating to adding at the README. In some cases, like with AWS Lambda functions, there may be some duplication of documentation. You may find that you are updating docs in the README, in the Lambda code file, and inline in the CloudFormation template. If you are reviewing changes, it's helpful to check any places that documentation may need to be updated to ensure that the repository stays in good shape.
```
Comment (gcampo88) [inline ci/CONTRIBUTING.MD]: "updating to adding at"?

File: ci/README.md [hunk 4]
```diff
@@ -1,84 +1,83 @@
...
-messages to GitHub and other places (Slack, email, etc).
+## Goals
 
-There is a third Lambda function that CodeBuild runs trigger to close out the
-process. That mainly entails updating the GitHub status, and sending more
-notifications to Slack/etc.
+In order to support the [practice](https://en.wikipedia.org/wiki/Continuous_integration) [of](https://aws.amazon.com/devops/continuous-integration/) **[continuous](https://www.visualstudio.com/learn/what-is-continuous-integration/) [integration](https://www.thoughtworks.com/continuous-integration)** for a large number of the apps and services we build, including all of our customer-facing services, we have built a platform to automate the test and build processes for each project.
 
-## Setup
+The high-level goals of the system are:
```
Comment (gcampo88) [inline ci/README.md]: I like these!

File: ci/README.md [hunk 5]
```diff
@@ -1,84 +1,83 @@
...
+- Push that archive to [Amazon S3](https://aws.amazon.com/s3/) (so that it is available later in the process)
+- Configure and trigger a build for the code in [AWS CodeBuild](https://aws.amazon.com/codebuild/)
+- Update the GitHub commit status for the code being tested
+- Send a notification to developers through channels like [Slack](https://slack.com/) so the process is highly visible
+
+#### CodeBuild
+
+When the previous step triggers a build, it does so by calling `startBuild` on a CodeBuild project that was setup by the `ci.yml` template specifically for running CI builds. All builds, regardless of which project or GitHub repository triggered them, are run through this one CodeBuild project.
+
+CodeBuild runs all builds in a Docker environment. Often it's the case that the code being built and tested is itself an application that runs on Docker. In such cases, those Docker containers and dealt with _inside_ of CodeBuild's native Docker build environment container (ie, Docker running inside Docker).
```
Comment (gcampo88) [inline ci/README.md]: "and dealt with" --> did you mean "are" ?

File: ci/README.md [hunk 6]
```diff
@@ -1,84 +1,83 @@
...
+- The location, in S3, of the code to be tested
+- Various environment variables that help with the rest of the CI process
+
+#### Build Callback
+
+Once a build run has completed, a third Lambda function is generally invoked to complete the CI process. Information about the original GitHub event, as well information about the results of the build run, are published to SNS, and the callback function handles those messages.
+
+Though the function handles some housekeeping tasks, such as updating the GitHub commit status and sending developer notifications, it's most important function is to publish information about deployable code artifacts that result from the builds. This only happens when such an artifact exists (ie not when building feature branches). Specifically, the function updates a [template configuration file](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/continuous-delivery-codepipeline-cfn-artifacts.html#w2ab2c13c15c15) that can be used by the [continuous deployment](https://github.com/PRX/Infrastructure/tree/master/cd) system to manage app deployments.
```
Comment (gcampo88) [inline ci/README.md]: its not it's

File: ci/README.md [hunk 7]
```diff
@@ -1,84 +1,83 @@
...
+If you opt in to using the common CI support code, the project must execute the `post_build.sh` script that is maintained in the Infrastructure repo. This is the code responsible for handling the common tasks associated with prepping apps for CD. If you choose to have this script publish the code resulting from the build process (which generally will be the case) the `PRX_CI_PUBLISH` environment variable must be set to the string value `true` by the time the post_build script runs. The `post_build` script also sends a message to an SNS topic, including information about the build process and any published code. There is a callback Lambda function that also runs as part of the CI process that can take these messages, and trigger updates to the CD system.
```
Comment (gcampo88) [inline ci/README.md]: and `PRX_CI_PUBLISH` will have been set to `'true'` by one of the lambdas if the project meets the CI requirements defined above, right? (This reads a little bit like developer should set that as part of buildspec)

**Comment by farski** [comment]
@gcampo88 think I covered all your edits

**Pull request merged**